### ISSUE TITLE
Thanks!

### ISSUE THREAD

**Issue by coderschoolreview** [khanhduyn/CoderSchool#1]
The goal of this assignment was to introduce you to 2 main concepts in Machine Learning: Data Pre-processing, and Classification. You learned how to query and clean data using the pandas library in Python, and built a simple Machine Learning Classifier based on the K Nearest Neighbors algorithm.

Looks like you know python quite well - your code could be cleaner, though. Your final submission has lots of commented code (and no actual comments) so it's a little difficult to follow. 

Good job on completing the main requirements of the assignment. In future weeks help the grader understand better comments or documentation sections in your code will be helpful. But great job with the plot and overall results.

**Comment by kunaljathal** [comment]
A couple of extra comments for assignment 1:
- The k value ranges are very wide and too high (that is why your graph looks a little crazy :)). We generally like to increase k in step size of 1, because kNN is quite sensitive to the value of k (egs: k_range = range(1, 20, 1))
- It's not clear how you arrived at rap vs jazz being the 2 genres that are most easily distinguishable, because you did not seem to try any other genres?

**Comment by coderschoolreview** [comment]
A couple of extra comments for assignment 1:

* The k value ranges are very wide and too high (that is why your graph looks a little crazy :)). We generally like to increase k in step size of 1, because kNN is quite sensitive to the value of k (egs: `k_range = range(1, 20, 1)`)
* It's not clear how you arrived at rap vs jazz being the 2 genres that are most easily distinguishable, because you did not seem to try any other genres?
But great job overall!

**Comment by coderschoolreview** [comment]
**Assignment 2**

The goal of this assignment was to introduce you to three new classification techniques and to understand how to select the best parameters and features for them. You learned how to use python built-in functions (GridSearchCV, SelectKBest, RFE, SelectFromModel) to try out new models (Support Vector Machines, Random Forests, and Logistic Regression) and test different permutations of parameter values and features, and analyze your results to help build better machine learning models.

*EXCELLENT* job! You are clearly very comfortable with programming and Python. 

Great work on the following:
* You used `moods` as features in addition to just audio, and you saw that your scores really increased. Well done!
* Plotted accuracy scores against increasing numbers of features, clearly showed the relationship between the two.
* Used the `Pipeline` object! Great work combining it to vary both `SelectKBest` and `LogisticRegression` values.
* In general you did nearly all the bonus stories and your work is very impressive!

Some suggestions:
* Good job commenting your code, you can add even more comments. Because you write quite a lot of code, comments really help other people (like your instructor :)) understand and follow your logic.
* You can try using `Pipeline` on even more combinations, egs: to vary both `RandomForestClassifier` parameter values along with `SelectFromModel` values. Here's some examples:
    * [RandomForestClassifier with SelectFromModel](https://stackoverflow.com/questions/35256876/ensuring-right-order-of-operations-in-random-forest-classification-in-scikit-lea)
    * [Extracting Features from Pipeline](https://stackoverflow.com/questions/35376293/extracting-selected-feature-names-from-scikit-pipeline)
* Looks like you got the following error:
`ConvergenceWarning: The max_iter was reached which means the coef_ did not converge`
You can try fixing this by increasing the `max_iter` parameter in `GridSearchCV` (the default is 100).
* Write a few sentences discussing your results. Such as:
    * Did all your classifiers improve with the best parameters? Why or why not?
    * Which classifiers seemed to be affected most by `GridSearchCV` ?
    * Which parameters made the biggest difference?

Overall, amazing work. Keep it up !!!

**Comment by coderschoolreview** [comment]
**Assignment 3**

The goal of this assignment was to introduce you to three new Natural Language Processing techniques, and to understand how to perform some basic sentiment analysis on song lyrics using these methods. You learned how to clean and prepare textual information for NLP, and then apply the following approaches: Bag Of Words, TF-IDF, and Doc2Vec. You used your prior knowledge of Python estimators, feature selection, and parameter optimization techniques to produce feature vectors from these NLP methods to make predictions on the moods of songs using their lyrics.

Amazing work. You do a great job on all the assignments and your submissions are always impressive.

What you did well:
* Very cool way of getting the rows where only happy or sad exist (XOR) !
* Awesome job exploring n-grams
* Great work in plotting the results of the n-grams vs accuracy
* Well done on experimenting with trying to use only English lyrics

Some notes:
* You plotted accuracy for the unbalanced dataset as well, and you used `accuracy_score` to measure this. However, this is a bit misleading -- you generally don't want to use `accuracy_score` when your data is unbalanced. It's better to use an [f1-score](http://scikit-learn.org/stable/modules/generated/sklearn.metrics.f1_score.html) or the [AUC](http://scikit-learn.org/stable/modules/generated/sklearn.metrics.auc.html). Check the following links for explanations on this:
    * https://www.quora.com/Why-is-it-that-I-get-a-better-accuracy-score-when-using-unbalanced-classes-with-a-Random-Forest-classifier
    * https://elitedatascience.com/imbalanced-classes

* Looks like you got a `ConvergenceWarning` for `GridSearchCV` on your `LogisticRegression` classifier; try bumping up `max_iter` even higher than 1000 and check if it works. However, note that in some cases, it is also possible that your data simply can't be fit by a logistic model.

Overall, excellent work!!
